lib/src/feature_discovery.dart

```diff
@@ -109,25 +110,45 @@ class DescribedFeatureOverlay extends StatefulWidget {
   final String title;
   final String description;
 
-  /// Called when the target is pressed.
-  final Function(VoidCallback onActionCompleted) doAction;
+  /// This is usually an [Icon].
+  /// The final tap target will already have a tap listener to finish each step.
+  ///
+  /// If you want to hit the tap target in integration tests, you should pass a [Key]
+  /// to this [Widget] instead of as the [Key] of [DescribedFeatureOverlay].
+  final Widget tapTarget;
 
-  /// Called just before the FeatureOverlay is displayed.
-  /// The function parameter is actually the callback that triggers the display of the overlay.
-  /// If not null, the callback MUST be called in order for the overlay to be displayed.
-  final Function(VoidCallback onActionCompleted) prepareAction;
   final Widget child;
   final ContentOrientation contentLocation;
   final bool enablePulsingAnimation;
 
-  /// Function to execute when the overlay is dismissed (when the user taps outside of it).
-  /// If not null, the callback MUST be called in order for the overlay to be dismissed.
-  final Function(VoidCallback onActionCompleted) onDismissAction;
+  /// Called just before the overlay is displayed.
+  /// This function needs to return a [bool], either from an `async` scope
+  /// or as a [Future].
+  ///
+  /// If this function returns `false`, this step will be marked complete
+  /// and therefore be skipped, i.e. it will not be opened.
+  /// In this case, we try to open the next step.
+  ///
+  /// When the [Future] finishes and evaluates to `true`, this step will be shown.
+  final Future<bool> Function() onOpen;
+
+  /// Called whenever the user taps outside the overlay area.
+  /// This function needs to return a [bool], either from an `async` scope
+  /// or as a [Future].
+  ///
+  /// If the function returns `false`, nothing happens. If it returns `true`,
+  /// all of the current steps are dismissed.
+  final Future<bool> Function() onDismiss;
+
+  /// Called when the tap target is tapped.
+  /// Whenever the [Future] this function returns is finished, the feature discovery
+  /// will continue and the next step will try to open after a closing animation.
+  final Future<void> Function() onTargetTap;
 
   const DescribedFeatureOverlay({
     Key key,
     @required this.featureId,
-    @required this.icon,
+    @required this.tapTarget,
     this.color,
     this.backgroundColor,
     this.iconColor,
```

Comment:
You have to delete this parameter, as it isn't used anymore.

Reply:
I am not exactly sure what you mean here. I renamed the `icon` parameter because it does not *need* to be an icon. The `tapTarget` parameter is used.

Reply:
Sorry, I was talking about the `iconColor` parameter!

Reply:
Oh, I see.